inmoose/pycombat/covariates.py

```diff
@@ -23,274 +23,372 @@
-    if ref_batch_idx is not None:
-        # the reference batch is not considered as a covariate
-        check[ref_batch_idx] = False
-    design = design[:, np.logical_not(check)]
-
-    logging.info(
-        f"Adjusting for {design.shape[1] - batchmod.shape[1]} covariate(s) or covariate level(s)"
-    )
-
-    # Check if the desigin is confounded
-    check_confounded_covariates(design, n_batch)
-
-    return (
-        design,
-        batchmod,
-        mod,
-        batches_ind,
-        n_batches,
-        n_batch,
-        n_sample,
-        ref_batch_idx,
-        batch,
-        remove_sample,
-    )
 
+    def __init__(self, counts, batch, covar_mod=None, ref_batch=None):
+        """
+        Arguments
+        ---------
+        counts : matrix
+            raw count matrix from genomic studies (dimensions gene x sample)
+        batch : array or list or :class:`~inmoose.utils.factor.Factor`
+            batch identifiers
+        covar_mod : list or matrix, optional
+            model matrix (dataframe, list or numpy array) for one or multiple
+            covariates to include in linear model (signal from these variables
+            are kept in data after adjustment). Covariates have to be
+            categorial, they can not be continious values (default: :code:`None`).
+        ref_batch : any
+            batch id of the batch to use as reference. Must be one of the element of
+            :code:`batch` (default: :code:`None`).
+        """
+        self.counts = counts
+        self.batch = asfactor(batch)
+        self.nan_batch = self.batch.isna().any()
+
+        nancounts = np.isnan(counts)
+        self.nan_genes = nancounts.any(axis=1).sum()
+        self.nan_samples = nancounts.any(axis=0).sum()
+
+        # number of batches
+        self.n_batch = self.batch.nlevels()
+        # number of samples in each batch
+        self.batch_composition = {
+            b: np.where(self.batch == b)[0] for b in self.batch.categories
+        }
+
+        if self.nan_batch:
+            self.batch_mod = None
+        else:
+            # batch design matrix
+            self.batch_mod = dmatrix("~0 + C(self.batch)")
+
+        # reference batch
+        if ref_batch is not None:
+            print(type(ref_batch))
+            if ref_batch not in self.batch.categories:
+                raise ValueError("Reference batch must identify one of the batches")
+            # ref_batch_idx is the index of the reference batch in batch.categories
+            self.ref_batch_idx = np.where(self.batch.categories == ref_batch)[0][0]
+            # update batchmod with reference
+            self.batch_mod[:, self.ref_batch_idx] = 1
+        else:
+            self.ref_batch_idx = None
+
+        if covar_mod is not None and not isinstance(covar_mod, DesignMatrix):
+            # convert to dataframe
+            if isinstance(covar_mod, (list, np.ndarray)):
+                covar_mod = pd.DataFrame(covar_mod)
```

Comment:
We check dimension for array and dataframe, should we check dimension for design matrix already in the good type?